### PR TITLE
Initial implementation of codesets (WIP)

### PR DESCRIPTION
**Pull request by jsuchome**
TODO:

- [x] how to solve gitea access from the server side
   UPDATE: server needs GITEA_URL, GITEA_USERNAME and GITEA_PASSWORD
- [x] non-push actions for codeset
   Get and List are there, querying the git server
- [ ] do we need real store for codeset, other then git repo?
- [ ] how to solve connection between codesets and pipelines (through tekton listeners)
- [x] add labels to repo, make them a lists or map

### CODE REVIEW

File: design/codeset.go
Comment (stefannica) [inline design/codeset.go:131]: I don't think the code path makes sense as an attribute of the REST API codeset resource, given that it only has meaning locally where the client is running.

File: design/codeset.go
Comment (stefannica) [inline design/codeset.go:137]: We should support associating not one but several labels with a single codeset. It would be even better if labels were modeled as dictionaries instead of lists of strings, because the dict syntax is better suited for matching purposes. For example, keys can represent classes of capabilities (e.g. library, app, predictor) that users and pipelines can filter and match to the values they support (e.g. `app: mlflow` and `library: sklearn` can be used to filter codesets that represent MLFlow projects using the scikit-learn library).

File: design/codeset.go
Comment (stefannica) [inline design/codeset.go:38]: We should use the same convention regarding naming for REST API paths, either all plural (runnables, codesets), or all singular.

File: design/codeset.go
Comment (jsuchome) [reply design/codeset.go:131]: Good point, it should be extra argument

File: design/codeset.go
Comment (jsuchome) [reply design/codeset.go:137]: Yes, I know this should be a list, I've just started simple

File: design/codeset.go
Comment (jsuchome) [reply design/codeset.go:38]: I think I have plural everywhere, just like with runnables. Or do you think it should be `GET("/codeset")` when retrieving single item?

File: pkg/core/codeset_client.go
Comment (jsuchome) [inline pkg/core/codeset_client.go:155]: I do not like this, it's directly copied from original carrier based code. I'd like to use gitea api for that

File: pkg/core/codeset_client.go
Comment (jsuchome) [inline pkg/core/codeset_client.go:24]: Note this fixed value. This is one listener created by fuseml tekton deployment. It might be necessary to always connect this to a new listener that executes specific pipeline...

File: pkg/core/codeset_client.go
Comment (stefannica) [inline pkg/core/codeset_client.go:32]: I wonder if the client indeed has to interact with the gitea API, or whether this is something that belongs on the server side, maybe as part of the codeset store implementation.

I would think that the CLI should only run those `git init ... git push` steps, because that's what a "regular" git user does anyway, while the rest (creating orgs, setting up webhooks and so on) should be done by the server. I mean, theoretically, these are two different things done through two separate APIs and even requiring different credentials/permissions: setting up the org/repository and pushing code to an existing repository.

What do you think about this separation of concerns: fuseml-core is a gitea admin user: it can access the gitea API and create/update/list projects (orgs) and create/update/list repositories (codesets), while the fuseml CLI acts as just a regular git client: it can only execute git operations (git clone, git push etc), but for everything else that requires access to the gitea API, it should go through the fuseml-core server ?

This separation plays well with another aspect: fuseml projects and codesets are FuseML specific "control plane" concepts, and the fuseml-core service is where they are centrally managed, so the CLI should call the server API for all operations that concern them. The only exception is the manipulation of codeset contents, which is just generic "payload" that can be transferred using another circuit - we can even call this the "data plane" - by using the git protocol that doesn't involve the fuseml-core API

It may just be me being heavily biased by my networking background, but this just feels like a better approach, what do you think ?

Comment (stefannica) [reply pkg/core/codeset_client.go:32]: Another argument I can bring in favor of this separation is that the client should be decoupled from the details of how the git server is implemented. It's gitea now, sure, but adding support for other git servers, or integrating FuseML with external git servers would be a lot cleaner if the implementation was managed server-side, while this is transparent for the client, given that it only "speaks" the generic git protocol expected to be implemented by any git server .

Comment (jsuchome) [reply pkg/core/codeset_client.go:32]: I agree that generally, whenever client is not required, it would be better to do the git operations from server.

So yes, as you mention: org manipulation and webhook creation could be done on server (as well as pulling for the sake of pipelines if they'd need it), while actual pushing the code needs to happen from client.

One thing I'm not sure about: how to pass the credentials? Do you mean server side should use the admin ones (like they are being used right now in this code) and client side should have specific user only ...
On little problem is that first such user would need to be created, so the `codeset register` would need to be split into 2 steps. But OK.

Next question is, how should I pass the credentials for the server side? Should it be part of the query? But that it means used would need to know it anyway. fuseml/carrier does it by checking the `secret` from gitea deployment, i.e. directly accessing kubernetes cluster. While I wanted to avoid this on client side, it could be fine on server...

As for gitea: so far I like the gitea API. As you can see, the client part are currently done with just executing git client though (but I'd like to change it to gitea too). But indeed, it might be good for the future to get independent.

File: design/codeset.go
Comment (jsuchome) [reply design/codeset.go]: This should be solved with latest commit

File: pkg/core/codeset_client.go
Comment (stefannica) [reply pkg/core/codeset_client.go:32]: I think the key to differentiating between what the client does directly and what the client does through the server is this: there are in fact two APIs exposed by the gitea server at work here, one is the gitea API that works _only_ with gitea (described in detail here: https://try.gitea.io/api/swagger#/), the other is the git protocol, that works with _any_ git server. Conversely, there are two types of client: the gitea client and the git client.

To keep the fuseml client implementation independent of the flavor of git server being used, only the git protocol should be used by the client when talking to the git server directly. This is sufficient for git repository operations like `git push` and `git pull` that are needed to transfer the codeset contents to/from the git server. All other operations that require access to the gitea API - things like managing (creating/deleting/configuring) organizations, repositories, credentials, webhooks etc. - should not be done by the client, but by the server instead. As mentioned in [my wiki PR](https://github.com/fuseml/wiki/pull/1), the fuseml client calls the FuseML server API to create a project (represented by a git organization) or a codeset (represented by a git repository), and it is the FuseML server which calls the gitea API to create those things.

> One thing I'm not sure about: how to pass the credentials? Do you mean server side should use the admin ones (like they are being used right now in this code) and client side should have specific user only ...
> On little problem is that first such user would need to be created, so the codeset register would need to be split into 2 steps. But OK.

Yes, when the server talks to the gitea API, it uses gitea's admin account (or another account with admin privileges). The client needs another set of credentials, one that is only required to authorize the client for `git push` operations in that repository (`git fetch/pull` being un-authenticated for "public" repositories anyway, same as github). To avoid having to implement the whole auth/authz feature just for this, I guess you could just create a hard-coded gitea user - let's call this the `gitops` user - and give it push access to all repositories. Either that, or configure codeset repositories so that you don't need to authorize users for push operations, if that's possible. You could even transfer the `gitops` user credentials back to the client in the POST response, if that makes things clearer from a client perspective.

> Next question is, how should I pass the credentials for the server side? Should it be part of the query? But that it means used would need to know it anyway. fuseml/carrier does it by checking the secret from gitea deployment, i.e. directly accessing kubernetes cluster. While I wanted to avoid this on client side, it could be fine on server...

I'm not sure I understand this question, but maybe my previous comments already help answer it. We can keep a single hard-coded user account to be used for clients when pushing code to any repository.

> As for gitea: so far I like the gitea API. As you can see, the client part are currently done with just executing git client though (but I'd like to change it to gitea too). But indeed, it might be good for the future to get independent.

I don't think the gitea API was ever meant to replace the git client commands. Yes, you can PUT files in a gitea repository using the gitea API, but it's not meant to be used instead of `git push`.

Comment (jsuchome) [reply pkg/core/codeset_client.go:32]: > I'm not sure I understand this question, but maybe my previous comments already help answer it. We can keep a > single hard-coded user account to be used for clients when pushing code to any repository.

I'm not talking about client side, but server side, which needs that gitea specific access. Currently I'm getting credentials from env variables on client side.  Should I 
- use different set of env variables on server side (for the admin user etc.)?
- pass the data through the REST API?
- fetch the gitea information from Kuberneres cluster

> I think the key to differentiating between what the client does ...

Yes, I agree that the client part does not need to be gitea specific as I already mentioned. But I'd keep it for next iteration.

> You could even transfer the gitops user credentials back to the client in the POST response, if that makes
> things clearer from a client perspective.

That was my original idea indeed. But maybe we can do with static user instead for now, as you suggest.

Comment (stefannica) [reply pkg/core/codeset_client.go:32]: > I'm not talking about client side, but server side, which needs that gitea specific access. Currently I'm getting credentials from env variables on client side.

The way I've seen this usually solved in cloud native applications is by having the fuseml-core service read the gitea admin user credentials from a configuration file, or better, from environment variables. How those env variables are passed onto the fuseml-core container depends on the method of installation. For example, if we had an uber-helm chart overseeing the configuration of all services that make up the fuseml app, I imagine the helm chart would have a global option holding the admin gitea password, which it would then use to configure both gitea itself and the fuseml-core service. Or it could generate the password, same as we did for SPR.

Comment (jsuchome) [reply pkg/core/codeset_client.go:32]: OK, I can settle on env variables that server reads for now. It could evolve later.

File: design/codeset.go
Comment (stefannica) [reply design/codeset.go:38]: Nevermind, for some reason I thought the runnable goa design was using `runnable` instead of `runnables`. I stand corrected.

File: pkg/core/codeset_client.go
Comment (jsuchome) [reply pkg/core/codeset_client.go]: Or rather git API, not gitea, as discussed elsewhere...

File: gen/http/openapi3.yaml
Comment (stefannica) [inline gen/http/openapi3.yaml:121]: `location` is still showing in the REST API as a POST op parameter. I think the only way to avoid this is to leave it out of the goa design spec completely and add it manually only as a CLI command line parameter, given that the `location` only makes sense on the client side.

File: pkg/core/client/git.go
Comment (stefannica) [inline pkg/core/client/git.go:44]: Just leaving this as a note here: there are two golang git libraries that could be used to achieve this, https://github.com/go-git/go-git and https://github.com/libgit2/git2go.

Comment (jsuchome) [reply pkg/core/client/git.go:44]: Yep, I've already noticed. But I will probably leave it for next PR.

**Review by stefannica — COMMENTED** [review]
Thanks for doing the client/server code separation. I have a few more suggestions regarding naming and code package structure, but overall I like how this is taking shape.

File: pkg/core/client/git.go
Comment (jsuchome) [reply pkg/core/client/git.go:4]: Yeah, that might be better indeed

File: pkg/core/client/git.go
Comment (jsuchome) [reply pkg/core/client/git.go:3]: Well, in the end, these values have to be the same, be it client or server. So I find it actually correct to share a config package

File: pkg/core/codeset_client.go
Comment (jsuchome) [reply pkg/core/codeset_client.go:1]: Yeah, the naming and locations are sometimes confusing, I'll have to think about it...

File: pkg/core/client/git.go
Comment (jsuchome) [reply pkg/core/client/git.go]: I might change the env var names to just GIT_* if you think it is better, but that's probably not your main point

Comment (stefannica) [reply pkg/core/client/git.go]: > Well, in the end, this value have to be the same, be it client or server. So I find it actually correct to share a config package

Conceptually, you wouldn't share a config package to be used by both client and server, because they're different entities and running as different processes. It just seems convenient in this case, because we're using hard-coded values for username/password in both client and server and they need to be coded somewhere common. This is more about importing in the client only what is needed by the client, and some of that stuff in the config package is only needed by the server.

Anyway, I agree this is more of a nitpick then something that needs changing, so you can leave it as is.

Comment (jsuchome) [reply pkg/core/client/git.go]: So if we get to the point when credentials for "fuseml user" are returned from the server so that client needs to use them explicitly, we would change this dependency indeed. Something for later then.

Comment (jsuchome) [reply pkg/core/client/git.go]: This would be https://github.com/fuseml/fuseml/issues/62

File: pkg/core/codeset_client.go
Comment (jsuchome) [reply pkg/core/codeset_client.go]: Check my latest commit

File: design/codeset.go
Comment (jsuchome) [reply design/codeset.go]: I change the labels type to list of strings

File: pkg/core/gitea/client.go
Comment (stefannica) [inline pkg/core/gitea/client.go:222]: One more reason to use a dictionary/map to represent labels :)

File: pkg/core/gitadmin.go
Comment (stefannica) [inline pkg/core/gitadmin.go:14]: This could be easily changed to fully apply the dependency inversion principle here. If you code this to be an interface instead of a struct, then you can instantiate a GiteaAdminClient object as a concrete implementation of GitAdmin and pass it to the codeset store to be used as a GitAdmin interface object. The way to tell if this principle works is when neither this file nor the codeset_store.go file has any direct dependencies to the GiteaAdminClient type.

**Review by stefannica — APPROVED** [review]
A few nitpicks and a suggestion about how you fully apply the dependency inversion principle with minimal changes on top of what you already did regarding code organization. I'll leave it up to you if you want to address them, otherwise I think this is ready to merge.

File: pkg/core/gitadmin.go
Comment (stefannica) [reply pkg/core/gitadmin.go:14]: Here's a link to a post concerning a dependency inversion problem that is very similar to ours: https://medium.com/@2hamed/dependency-isolation-in-go-afab00b5e4c9

File: design/codeset.go
Comment (jsuchome) [reply design/codeset.go]: Yep, let's leave it for the time we have to solve it.

File: go.mod
Comment (jsuchome) [reply go.mod:7]: Honestly, I do not know. I forgot to check closely, likely you are right.

File: pkg/core/gitea/client.go
Comment (jsuchome) [reply pkg/core/gitea/client.go:222]: Well, maybe. The problem here is not with _our_ structure, but with the gitea API that returns `[]string` .

(Than you might want to say, let's not use git for labels - and I guess we'll see if we need some extra store...)

File: pkg/client/git.go
Comment (jsuchome) [reply pkg/client/git.go:59]: fixed

File: pkg/core/codeset_store.go
Comment (stefannica) [inline pkg/core/codeset_store.go:7]: For the dependency inversion to be fully effective, I would say you shouldn't need to add any dependency here to the gitea package. The quote goes: _High-level modules should not depend on low-level modules. Both should depend on abstractions (e.g. interfaces)_. In this case I think that means the codeset store shouldn't depend on anything from `core/gitea`, it should only import the interface from `core/gitadmin`.

Comment (jsuchome) [reply pkg/core/codeset_store.go:7]: I don't think so. There needs to be some code that actually imports the file where the implementation is happening. To quote some other part of that article,

 _`main` is the only package which needs to be aware of those 2 components_. 

(`codeset_store.go` is our `main` here)

Comment (stefannica) [reply pkg/core/codeset_store.go:7]: I think that was meant literally, and `fuseml_core/main` is our main package :)

Comment (jsuchome) [reply pkg/core/codeset_store.go:7]: No, I do not think so, in that example it's `main` package which calls the storage and foo parts, that's why I'm saying `codeset_store.go` is our `main` compared to this example

Comment (jsuchome) [reply pkg/core/codeset_store.go:7]: My point is, _someone_ needs to instantiate NewGiteaAdminClient before it can be passed to interface. And you just can't do it without some imports.

Anyway, maybe I'm missing something, this golang-specific concept is still confusing to me...

**Comment by jsuchome** [comment]
Let's leave remaining work for follow-ups

**Pull request merged**
